### PR TITLE
Improve PKCS11 unit test, fix findings in implementation

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: src/common/pkcs11/pkcs11.cpp [hunk 1]
```diff
@@ -1102,7 +1102,7 @@ Error Utils::ImportCertificate(const Array<uint8_t>& id, const String& label, co
     CK_BBOOL            trueVal      = CK_TRUE;
     CK_BBOOL            falseVal     = CK_FALSE;
 
-    StaticArray<uint8_t, crypto::cSerialNumSize> serialNum;
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:1105]: Commit message too generic: what exactly fixed?

File: src/common/pkcs11/pkcs11.cpp [hunk 2]
```diff
@@ -1102,7 +1102,7 @@ Error Utils::ImportCertificate(const Array<uint8_t>& id, const String& label, co
...
+    StaticArray<uint8_t, crypto::cSerialNumSize * 2> serialNum;
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:1102]: One byte of serial is encoded to 2 bytes of asn1?

File: include/aos/common/crypto/mbedtls/cryptoprovider.hpp [hunk 3]
```diff
@@ -133,7 +133,8 @@ class MbedTLSCryptoProvider : public aos::crypto::x509::ProviderItf {
 private:
     aos::Error ParseX509Certs(mbedtls_x509_crt* currentCrt, aos::crypto::x509::Certificate& cert);
     aos::Error GetX509CertExtensions(aos::crypto::x509::Certificate& cert, mbedtls_x509_crt* crt);
-    void       GetX509CertData(aos::crypto::x509::Certificate& cert, mbedtls_x509_crt* crt);
+    aos::Error GetX509CertData(aos::crypto::x509::Certificate& cert, mbedtls_x509_crt* crt);
+    aos::Error ConvertTime(const mbedtls_x509_time& src, aos::Time& dst);
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/cryptoprovider.hpp:137]: Commit message is too generic. What exactly is fixed?

File: tests/common/src/pkcs11/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,99 @@
+#
+# Copyright (C) 2023 Renesas Electronics Corporation.
```
Comment (al1img) [inline tests/common/src/pkcs11/CMakeLists.txt:2]: 2024

File: tests/common/src/pkcs11/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+# Create PKCS11 test certificates
+# ######################################################################################################################
+
+set(CERTDIR ${CMAKE_CURRENT_BINARY_DIR}/certificates)
+
+file(MAKE_DIRECTORY ${CERTDIR})
+
+message("\nGenerating PKCS11 test certificates...")
+
+message("  1) Create a Certificate Authority private key...")
```
Comment (al1img) [inline tests/common/src/pkcs11/CMakeLists.txt:41]: remove numbering: 1), 2) - when you have to add or remove something you should always reorder this.

File: tests/common/src/pkcs11/pkcs11_test.cpp [hunk 6]
```diff
@@ -474,5 +474,11 @@ TEST_F(PKCS11Test, FindCertificateChain)
     ASSERT_EQ((*chain)[1].mIssuer, caCert.mIssuer);
 }
 
+TEST_F(PKCS11Test, PKCS11RSAPrivateKeySign)
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11_test.cpp:598]: It is empty. Do we need it?

File: include/aos/common/crypto.hpp [hunk 7]
```diff
@@ -176,11 +176,11 @@ class PublicKeyItf {
  */
 class HashType {
 public:
-    enum class Enum { eSHA1, eSHA224, eSHA256, eSHA384, eSHA512 };
+    enum class Enum { eSHA1, eSHA224, eSHA256, eSHA384, eSHA512, eNone };
```
Comment (al1img) [inline include/aos/common/crypto.hpp:179]: Could we put `eNone` to the beginning?

File: tests/common/src/pkcs11/pkcs11_test.cpp [hunk 8]
```diff
@@ -162,6 +167,80 @@ Error ReadFile(const char* filename, Array<uint8_t>& array)
     return ErrorEnum::eNone;
 }
 
+static void importRSAPublicKey(const crypto::RSAPublicKey& rsaKey, mbedtls_pk_context& ctx)
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11_test.cpp:170]: From capital letter.

File: tests/common/src/pkcs11/pkcs11_test.cpp [hunk 9]
```diff
@@ -162,6 +167,80 @@ Error ReadFile(const char* filename, Array<uint8_t>& array)
...
+
+    mbedtls_rsa_context* rsaCtx = mbedtls_pk_rsa(ctx);
+
+    ASSERT_EQ(mbedtls_rsa_import_raw(rsaCtx, n.Get(), n.Size(), nullptr, 0, nullptr, 0, nullptr, 0, nullptr, 0), 0);
+    ASSERT_EQ(mbedtls_rsa_import_raw(rsaCtx, nullptr, 0, nullptr, 0, nullptr, 0, nullptr, 0, e.Get(), e.Size()), 0);
+    ASSERT_EQ(mbedtls_rsa_complete(rsaCtx), 0);
+    ASSERT_EQ(mbedtls_rsa_check_pubkey(rsaCtx), 0);
+}
+
+static void importECDSAPublicKey(const crypto::ECDSAPublicKey& ecdsaKey, mbedtls_pk_context& ctx)
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11_test.cpp:185]: ditto

File: include/aos/common/crypto.hpp (hunk 7 above)
Comment (mykola-kobets-epam) [reply include/aos/common/crypto.hpp]: yes, fixed

File: tests/common/src/pkcs11/pkcs11_test.cpp (hunk 8 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11_test.cpp]: fixed

File: tests/common/src/pkcs11/pkcs11_test.cpp (hunk 9 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11_test.cpp]: fixed

File: tests/common/src/pkcs11/pkcs11_test.cpp (hunk 6 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11_test.cpp:598]: squashed with the next commit

File: include/aos/common/crypto/mbedtls/cryptoprovider.hpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply include/aos/common/crypto/mbedtls/cryptoprovider.hpp:137]: PublicKey export is supposed to be done in a separate PR due to complexity

File: src/common/pkcs11/pkcs11.cpp [hunk 10]
```diff
@@ -674,19 +674,16 @@ Error SessionContext::Decrypt(
         return err;
     }
 
-    CK_ULONG resultSize = 0;
+    // SoftHSM doesn't provide a precise size with Decrypt(data, nullptr, &resultSize) call.
+    CK_ULONG resultSize = result.MaxSize();
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:678]: empty line after var definition

File: tests/common/src/crypto_test.cpp [hunk 11]
```diff
@@ -758,8 +758,14 @@ TEST(CryptoTest, ASN1EncodeBigInt)
 
     ASSERT_EQ(provider.ASN1EncodeBigInt(inputBigInt, asn1Value), aos::ErrorEnum::eNone);
 
-    std::vector<uint8_t>       actual       = {asn1Value.begin(), asn1Value.end()};
-    const std::vector<uint8_t> expectedASN1 = {0x2, 0x8, 0x17, 0xad, 0x4f, 0x60, 0x5c, 0xda, 0xe7, 0x9e};
+    std::vector<uint8_t> actual = {asn1Value.begin(), asn1Value.end()};
+
+    // TODO:
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply tests/common/src/crypto_test.cpp:763]: ok

File: tests/common/src/pkcs11/pkcs11_test.cpp [hunk 12]
```diff
@@ -140,7 +141,11 @@ RetWithError<UniquePtr<SessionContext>> PKCS11Test::OpenUserSession(bool login)
     return session;
 }
 
-Error ReadFile(const char* filename, Array<uint8_t>& array)
+/***********************************************************************************************************************
+ * Static
+ **********************************************************************************************************************/
+
+static Error ReadFile(const char* filename, Array<uint8_t>& array)
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11_test.cpp:148]: fixed

File: src/common/pkcs11/pkcs11.cpp (hunk 10 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:678]: added

**Review by MykolaSolyanko — APPROVED** [review]

**Pull request merged**